lib/babe/epoch_handler.go

```diff
@@ -75,83 +75,27 @@ func (h *epochHandler) run(ctx context.Context, errCh chan<- error) {
 		return
 	}
 
-	// for each slot we're handling, create a timer that will fire when it starts
-	// we create timers only for slots where we're authoring
-	authoringSlots := getAuthoringSlots(h.slotToPreRuntimeDigest)
-
-	type slotWithTimer struct {
-		startTime time.Time
-		timer     *time.Timer
-		slotNum   uint64
-	}
-
-	slotTimeTimers := make([]*slotWithTimer, 0, len(authoringSlots))
-	for _, authoringSlot := range authoringSlots {
-		if authoringSlot < currSlot {
-			// ignore slots already passed
-			continue
-		}
-
-		startTime := getSlotStartTime(authoringSlot, h.constants.slotDuration)
-		waitTime := time.Until(startTime)
-		timer := time.NewTimer(waitTime)
-
-		slotTimeTimers = append(slotTimeTimers, &slotWithTimer{
-			timer:     timer,
-			slotNum:   authoringSlot,
-			startTime: startTime,
-		})
-
-		logger.Debugf("start time of slot %d: %v", authoringSlot, startTime)
-	}
-
-	logger.Debugf("authoring in %d slots in epoch %d", len(slotTimeTimers), h.epochNumber)
-
-	for _, swt := range slotTimeTimers {
-		logger.Debugf("waiting for next authoring slot %d", swt.slotNum)
+	logger.Debugf("authoring in %d slots in epoch %d", len(h.slotToPreRuntimeDigest), h.epochNumber)
 
+	for {
 		select {
 		case <-ctx.Done():
-			for _, swt := range slotTimeTimers {
-				swt.timer.Stop()
-			}
+			errCh <- nil
 			return
-		case <-swt.timer.C:
-			// we must do a time correction as the slot timer sometimes is triggered
-			// before the time defined in the constructor due to an inconsistency
-			// of the language -> https://github.com/golang/go/issues/17696
-
-			diff := time.Since(swt.startTime)
-			if diff < 0 {
-				time.Sleep(-diff)
-			}
-
-			if _, has := h.slotToPreRuntimeDigest[swt.slotNum]; !has {
-				// this should never happen
-				panic(fmt.Sprintf("no VRF proof for authoring slot! slot=%d", swt.slotNum))
-			}
-
-			currentSlot := Slot{
-				start:    swt.startTime,
-				duration: h.constants.slotDuration,
-				number:   swt.slotNum,
-			}
-			err := h.handleSlot(h.epochNumber, currentSlot, h.epochData.authorityIndex, h.slotToPreRuntimeDigest[swt.slotNum])
-			if err != nil {
-				logger.Warnf("failed to handle slot %d: %s", swt.slotNum, err)
-				continue
-			}
+		default:
 		}
-	}
-}
 
-// getAuthoringSlots returns an ordered slice of slot numbers where we can author blocks,
-// based on the given VRF output and proof map.
-func getAuthoringSlots(slotToPreRuntimeDigest map[uint64]*types.PreRuntimeDigest) []uint64 {
-	authoringSlots := maps.Keys(slotToPreRuntimeDigest)
-	sort.Slice(authoringSlots, func(i, j int) bool {
-		return authoringSlots[i] < authoringSlots[j]
-	})
+		currentSlot := h.slotHandler.waitForNextSlot()
```

Comment:
This is a blocking call, so the `epochHandler` will not be able to receive the cancellation signal.  I think we need to be able to abandon this call if the cancellation signal comes in.